sire/advanced.md

```diff
@@ -26,70 +26,104 @@ This line is the type signature:
 > Bar > Bar > Nat > Bit
 ```
 
-First parameter is a bar ("needle"), second parameter is a bar ("haystack"), third parameter is a nat (offset) and the function returns a bit (boolean).
+First parameter is a bar (`needle`), second parameter is a bar (`haystack`), third parameter is a nat (`off`, for offset) and the function returns a bit (boolean).
 
 ## Lambdas
 
 - `&`  Anonymous lambda
 - `?`  Named lambda
-- `??` Named and pinned lambda (Pinning has to do with helping with memory layout in persistence. Don't worry about it too much for now, but you'll come across it in source files)
+- `??` Named and pinned lambda (Pinning has to do with memory layout. Don't worry about it too much for now, but you'll come across it in source files)
 
-## Col macro
+For example, here's `&` used infix:
 
-"Col" as in "colon"
+```sire
+| map x&(add x | mul x 2) [0 1 2]
+[0 3 6]
+```
+
+And here it's used prefix:
 
 ```sire
-: (a b c) < foo x y
-; the rest of the computation below
+| foreach [0 1 2]
+& x
+| add x | mul x 2
+[0 3 6]
 ```
-- `foo x y` is the function call. It accepts another function as an argument - the callback below:
-- `(a b c)` `a b c` are the arguments to the callback
-- The main function and callback can have any arguments - these were chosen arbitrarily for this example
 
-We definitely need an example to better understand this. We'll use `gen` for the example (`gen` was covered in the [standard library](/sire/standard-library.md))
+Both `map` and `foreach` do the same thing, they just accept their arguments in different orders. `map` takes a function and then a row, while `foreach` takes a row and then a function.
+
+## Col macro
+
+Using `&` prefix in the way we just did is quite uncommon in practice. We have a macro called `:` (pronounced "col") which does the same thing in one line less:
 
 ```sire
-; we saw gen used like this:
-gen 10 | mul 2
-[0 2 4 6 8 10 12 14 16 18]
+: x < foreach [0 1 2]
+| add x | mul x 2
+[0 3 6]
+```
 
-; Here it is with col:
+This macro-expands into exactly the same code as the version with `&`. The only difference is that it feels more like an assignment.
 
-: index < gen 5
-| mul index 2
-[0 2 4 6 8]
+It's worth thinking about the shape of this a little bit more. If we had types, the type of `foreach` would be `Row a -> (a -> b) -> Row b`.
+So `foreach` takes two arguments, a row and a function. We can clearly see this in the version that uses `&`: the first argument is the row `[0 1 2]` and the second argument spans lines 2 and 3.
+But when we write it using the col macro, it *feels* as if we're only giving `foreach` a single argument, the row. And then somehow it *feels* as if `foreach` *returns* an `x` which we can use in the "loop body". In reality there is no loop body, and we're still passing a function as a second argument. But in many situations, this assignment-like syntax ends up communicating the intent much more clearly.
```

Comment:
This is the best explanation of the col macro i've seen so far. I think you may have said this aloud to me before, but reading it is even better.